From b29b06d9c406b9c2721742dfb14d7806973e894d Mon Sep 17 00:00:00 2001
From: Jan Engelhardt <jengelh@inai.de>
Date: Wed, 14 Oct 2020 12:26:10 +0200
Subject: [PATCH 53/73] base: fix nullptr deref around GetTaskExecutorForTraits

When trknotify.cpp tries to post a UI task, some executors
are still nullptr.
---
 base/task/post_task.cc                        | 28 +++++++++++++------
 .../scheduler/browser_task_executor.cc        | 13 ++++++++-
 2 files changed, 32 insertions(+), 9 deletions(-)

diff --git base/task/post_task.cc base/task/post_task.cc
index 434c7c3f126c8..9ecba7b3f8c21 100644
--- base/task/post_task.cc
+++ base/task/post_task.cc
@@ -72,8 +72,10 @@ bool PostDelayedTask(const Location& from_here,
                      const TaskTraits& traits,
                      OnceClosure task,
                      TimeDelta delay) {
-  return GetTaskExecutorForTraits(traits)->PostDelayedTask(
-      from_here, traits, std::move(task), delay);
+	auto e = GetTaskExecutorForTraits(traits);
+	if (e == nullptr)
+		return false;
+	return e->PostDelayedTask(from_here, traits, std::move(task), delay);
 }
 
 bool PostTaskAndReply(const Location& from_here,
@@ -85,27 +87,37 @@ bool PostTaskAndReply(const Location& from_here,
 }
 
 scoped_refptr<TaskRunner> CreateTaskRunner(const TaskTraits& traits) {
-  return GetTaskExecutorForTraits(traits)->CreateTaskRunner(traits);
+	auto e = GetTaskExecutorForTraits(traits);
+	if (e == nullptr)
+		return nullptr;
+	return e->CreateTaskRunner(traits);
 }
 
 scoped_refptr<SequencedTaskRunner> CreateSequencedTaskRunner(
     const TaskTraits& traits) {
-  return GetTaskExecutorForTraits(traits)->CreateSequencedTaskRunner(traits);
+	auto e = GetTaskExecutorForTraits(traits);
+	if (e == nullptr)
+		return nullptr;
+	return e->CreateSequencedTaskRunner(traits);
 }
 
 scoped_refptr<SingleThreadTaskRunner> CreateSingleThreadTaskRunner(
     const TaskTraits& traits,
     SingleThreadTaskRunnerThreadMode thread_mode) {
-  return GetTaskExecutorForTraits(traits)->CreateSingleThreadTaskRunner(
-      traits, thread_mode);
+	auto e = GetTaskExecutorForTraits(traits);
+	if (e == nullptr)
+		return nullptr;
+	return e->CreateSingleThreadTaskRunner(traits, thread_mode);
 }
 
 #if defined(OS_WIN)
 scoped_refptr<SingleThreadTaskRunner> CreateCOMSTATaskRunner(
     const TaskTraits& traits,
     SingleThreadTaskRunnerThreadMode thread_mode) {
-  return GetTaskExecutorForTraits(traits)->CreateCOMSTATaskRunner(traits,
-                                                                  thread_mode);
+	auto e = GetTaskExecutorForTraits(traits);
+	if (e == nullptr)
+		return nullptr;
+	return e->CreateCOMSTATaskRunner(traits, thread_mode);
 }
 #endif  // defined(OS_WIN)
 
diff --git content/browser/scheduler/browser_task_executor.cc content/browser/scheduler/browser_task_executor.cc
index 82801d316d76c..8f594bfe26142 100644
--- content/browser/scheduler/browser_task_executor.cc
+++ content/browser/scheduler/browser_task_executor.cc
@@ -260,7 +260,8 @@ void BrowserTaskExecutor::Shutdown() {
 void BrowserTaskExecutor::RunAllPendingTasksOnThreadForTesting(
     BrowserThread::ID identifier) {
   DCHECK(Get());
-
+  if (Get() == nullptr)
+    return;
   base::RunLoop run_loop(base::RunLoop::Type::kNestableTasksAllowed);
 
   switch (identifier) {
@@ -282,6 +283,8 @@ void BrowserTaskExecutor::RunAllPendingTasksOnThreadForTesting(
 
 // static
 void BrowserTaskExecutor::EnableAllQueues() {
+  if (Get() == nullptr)
+    return;
   Get()->browser_ui_thread_handle_->EnableAllQueues();
   Get()->browser_io_thread_handle_->EnableAllQueues();
 }
@@ -289,21 +292,29 @@ void BrowserTaskExecutor::EnableAllQueues() {
 // static
 scoped_refptr<base::SingleThreadTaskRunner>
 BrowserTaskExecutor::GetUIThreadTaskRunner(const BrowserTaskTraits& traits) {
+  if (Get() == nullptr)
+    return nullptr;
   return Get()->GetTaskRunner(BrowserThread::UI, traits);
 }
 
 // static
 scoped_refptr<base::SingleThreadTaskRunner>
 BrowserTaskExecutor::GetIOThreadTaskRunner(const BrowserTaskTraits& traits) {
+  if (Get() == nullptr)
+    return nullptr;
   return Get()->GetTaskRunner(BrowserThread::IO, traits);
 }
 
 // static
 void BrowserTaskExecutor::InitializeIOThread() {
+  if (Get() == nullptr)
+    return;
   Get()->browser_io_thread_handle_->EnableAllExceptBestEffortQueues();
 }
 
 std::unique_ptr<BrowserProcessSubThread> BrowserTaskExecutor::CreateIOThread() {
+  if (Get() == nullptr)
+    return nullptr;
   DCHECK(Get()->io_thread_executor_);
 
   std::unique_ptr<BrowserIOThreadDelegate> browser_io_thread_delegate =
-- 
2.32.0

